polaris/hub/client.py

```diff
@@ -662,18 +720,193 @@ def upload_benchmark(
         benchmark_json["access"] = access
 
-        url = f"/benchmark/{benchmark.owner}/{benchmark.name}"
+        path_params = "v1/benchmark" if artifact_type == ArtifactType.STANDARD.value else "v2/competition"
+        url = f"/{path_params}/{benchmark.owner}/{benchmark.name}"
         response = self._base_request_to_hub(url=url, method="PUT", json=benchmark_json)
 
         logger.success(
-            "Your benchmark has been successfully uploaded to the Hub. "
-            f"View it here: {urljoin(self.settings.hub_url, f'benchmarks/{benchmark.owner}/{benchmark.name}')}"
+            f"Your {artifact_type} has been successfully uploaded to the Hub. "
+            f"View it here: {urljoin(self.settings.hub_url, f'{artifact_type}/{benchmark.owner}/{benchmark.name}')}"
         )
         return response
+
+    def upload_competition(
+        self,
+        competition: CompetitionSpecification,
+        timeout: TimeoutTypes = (10, 200),
+        owner: Optional[Union[HubOwner, str]] = None,
+        if_exists: ZarrConflictResolution = "replace",
+    ) -> Dict[str, Response]:
+        """Upload a competition to the Polaris Hub.
+
+        Args:
+            competition: The competition to upload.
+            timeout: Request timeout values. User can modify the value when uploading large dataset as needed.
+                This can be a single value with the timeout in seconds for all IO operations, or a more granular
+                tuple with (connect_timeout, write_timeout). The type of the the timout parameter comes from `httpx`.
+                Since datasets can get large, it might be needed to increase the write timeout for larger datasets.
+                See also: https://www.python-httpx.org/advanced/#timeout-configuration
+            owner: Which Hub user or organization owns the artifact. Takes precedence over `dataset.owner`.
+            if_exists: Action for handling existing files in the Zarr archive. Options are 'raise' to throw
+                an error, 'replace' to overwrite, or 'skip' to proceed without altering the existing files.
+
+        Returns:
+            A object containing responses from the Polaris Hub regarding both the dataset and benchmark upload.
+        """
+        ACCESS = "private"
+
+        if competition.dataset.zarr_root is not None:
+            logger.info("Zarr-based competitions are not supported at this time")
+            return
+
+        # Upload competition dataset
+        dataset_response = self._upload_dataset(
+            competition.dataset, ArtifactType.COMPETITION.value, ACCESS, timeout, owner, if_exists
+        )
+        # Upload competition benchmark
+        competition_response = self._upload_benchmark(
+            competition, ArtifactType.COMPETITION.value, ACCESS, owner
+        )
+
+        return {
+            "dataset_response": dataset_response,
+            "competition_response": competition_response,
+        }
```

Comment:
**Comment**
It's a little late in the game, but I think this illustrates what I was saying early in the design process, where a competition isn't `dataset + benchmark`, it's `competition`. It _looks_ like it's two things (because we have two similar things, data wise), but it doesn't _behave_ like two things. It's one thing.

Unfortunately, that means we end up with some leaking abstraction here, where if the second request to the Hub fails, the user is stuck with half a competition and possibly can't recover.

Having a single API endpoint to upload all the metadata at once, and then following up with the blob content upload to the bucket, would provide a better user experience.

Reply:
I brought this up with Andrew yesterday -- you're right that it's possible to end up in a situation where the dataset and benchmark parts of a competition are out of sync and it's not fixable (because if re-try it will say the dataset already exists even though there's no benchmark (competition) with it. I think we should think through how to handle the case where a dataset gets orphaned and can't be used (or the competition gets orphaned, e.g. if the dataset pre-processing fails).

Reply:
Yeah this could be troublesome. It's also a difficult problem to completely fix/overcome because we at the least have two independent writes happening where the name must be unique (1 for the R2 blob, 1 transaction for dataset and competition records, assuming we bundle them in a transaction).

We'll have to put one of these in front of the other, and if it fails, the user still won't be able to retry with the same name. Definitely worth a chat as Kira mentioned.